### DISCUSSION TITLE
When is `CompilationProvider` triggered if used as a ValueProvider for Source Generation?

### DISCUSSION THREAD

**Discussion by thomhurst** [dotnet/roslyn#75988] [Q&A]
If I was to do something like this, when/how often would it trigger?

```
context.RegisterSourceOutput(context.CompilationProvider, (productionContext, compilation) => GenerateCode(productionContext, compilation));
```

I want to run the source generator once, since I just want to generate some metadata about the project, so I don't need to inspect the code and inspect SyntaxNodes.
However as part of my generator, I do need access to the `Compilation` object. So I was assuming the CompilationProvider would make the most sense? I have tried CompilationProvider, but my generated classes don't seen to be available in the IDE.

If I emit them in the csproj, I can see they get generated on a build, but does that mean they're not available before that?
I need them available to users before they build, like keystroke/syntax node source generators would be.

**Comment by 333fred** [comment]
> when/how often would it trigger?

Every keystroke. Do not do this. Make an equatable model to at least dampen the impact.

**Reply by thomhurst** [reply]
Yeah I did do this. The ide still refuses to show it in intellisense 🤔

**Reply by Dreamescaper** [reply]
There are multiple bugs related to Balanced source generator mode. Frankly, I don't get why it's enabled by default if it doesn't work.
(https://github.com/dotnet/roslyn/issues/74032, https://github.com/dotnet/roslyn/issues/75586, https://github.com/dotnet/roslyn/issues/74601)
Try selecting Automatic (Options -> Text Editor -> C# -> Advanced).

**Reply by CyrusNajmabadi** [reply]
74601 is unrelated to balanced mode. 
74032 is about a button we added to explicitly dump cached generator state (regardless of balanced or automatic mode) in case an incremental generator stores invalid intermediary data. 

75586 seems like an issue there. But I don't repro it all. So it's possible something else is breaking the workspace on your machine.

**Reply by CyrusNajmabadi** [reply]
> Frankly, I don't get why it's enabled by default if it doesn't work.

It's on by default for all users and we've only gotten a tiny handful of issues (far fewer reports than all the reports we got about automatic mode). Data indicates that the problems are being hit by a tiny minority. This likely indicates something odd or uncommon about some setup that we haven't been able to track down yet. so we've kept it enabled by default for the *massive* perf and stability gains the rest of the community has gotten. 

If you want automatic mode, feel free to turn it on. That's why we have a switch. But turning on SGs in automatic mode originally was terrible, with tanking perf and stability across tons of users. That's why we flipped things.

**Reply by Dreamescaper** [reply]
> 74032 is about a button we added to explicitly dump cached generator state (regardless of balanced or automatic mode) in case an incremental generator stores invalid intermediary data.

Maybe it's not related from technical side, but it's related from use-case perspective. SG doesn't update the file due to 75586, therefore I'm trying to force it to regenerate by pressing that button, and I get VS completely broken until restart due to 74032. So it's probably not surprising that Balanced mode feels half-baked at least in my particular case.

Sorry, if I might sound overly aggressive. Sometimes I am confused by the priorities, but I do realize I probably don't see the whole picture.

In this case, I think that most people simply don't care about such issues, and fine with the source generator working only on build. Very small amount of people use source generators explicitly in day-to-day work, and a tiny fraction of those develop SGs. Contrary to simply writing regular code, which automatic mode heavily affected.

**Reply by CyrusNajmabadi** [reply]
> In this case, I think that most people simply don't care about such issues

The number of reports we got about performance was very high. Many people had serious and significant slowdowns that made vs much worse (or even unusable). 

Again, if you are having issues with balanced mode, then go back to automatic mode. It's an option for that reason, and we have no plans to remove the option.

**Reply by thomhurst** [reply]
Is this a Visual Studio thing? I'm using Rider

**Reply by CyrusNajmabadi** [reply]
Yes.  Roslyn is the impl of C# that operates in VS.  Rider is entirely different and issues you're running into should go to them.

**Reply by thomhurst** [reply]
I think it's a rider IDE issue I'm experiencing actually. My app compiles and builds, but Rider says source generated types are missing. But they're clearly not if it runs fine.
